Created pluggable authentication fetch function

EliteScientist opened the pull request:
Created pluggable auth fetch function where an authenticator can be passed via the options to support pre-authentication, authentication fault detection, and authentication.

Commit:
Created pluggable authentication fetch function

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+import { finCsrfFetch } from './finCsrfFetch'
```
Needs copyright

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+		}
+
+		// Pipe request to the csrf fetch function
+		resp = await finCsrfFetch(resource, options)
```
I don't think we should have a hard coded dependency on `finCsrfFetch`. One should be able to pass in an alternative fetch implementation. Then everything can be plugged together into a chain for use.

That being said, if a fetch isn't specified in the arguments then it can default back to `finCsrfFetch`.

EliteScientist replied:
I agree. I kept thinking about this one also.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+		const authenticator = options.authenticator as RequestAuthenticator
+
+		// Pre authenticate if suppprted by authenticator
+		if (authenticator.preauthenticate) {
```
Prefer camel case `preAuthenticate`.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+		}
+	} else {
+		// If request doesn't contain an authenticator, continue has normal
+		resp = await finCsrfFetch(resource, options)
```
The fetch is duplicated at both line 84 and 46 for no reason. If there's a fetch and it's successful then it shouldn't fetch again. 

I think you can just remove the else statement completely.

EliteScientist replied:
46 is needed and only called if there is an authenticator passed.

Line 84 is called if there are no options and no authenticator passed, this functions just like fetch without an authenticator.

garethj2 replied:
I see now - my bad. I find the code review format in github difficult to follow.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+	 * @param request to pre authenticate
+	 * @returns Pre authenticated Request Object
+	 */
+	preauthenticate?: (request: RequestInfo) => Promise<Request>
```
I think the original options should be passed to both `preAuthenticate` and `authenticate`.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -0,0 +1,144 @@
+			}
+
+			// Attempt to authenticate until' generator has reached the max try threshold
+			for await (const didAuthenticate of tryAuthentication()) {
```
I see why you're using a generator here but I think a simple for loop without the generator would be easier to follow...

        // Check if response is an authentication fault
	    if (!(await authenticator.isAuthenticated?.(resp))) {
			const maxTries = authenticator.maxTries ?? 3
			let authSuccessful = false

			for (let i = 0; i < maxTries; ++i) {
				const result = await authenticator.authenticate(
					resp as Response
				)

				if (result) {
					authSuccessful = true
					break
				}
			}

			// If we reached the max try threshold and authentication was still unsuccessful
			// throw an exception
			if (!authSuccessful) {
				throw new Error('Authentication Failed')
			}

			// Replay request
			resp = await finCsrfFetch(resource, options)
		}

EliteScientist replied:
Ahh. Yes, I coded this as if it was using the forEach callback where it doesn't pause during an await. I agree it will be easier to follow that way with the for loop

Commit:
PR Changes:
 - Add Copyright header
 - Make fetch function pluggable
 - camel case preAuthenticate
 - pass request options to authenticate and preAuthenticate
 - change for..of generator loop to for..i loop

Commit:
change comment copy

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -77,10 +76,11 @@ export async function finAuthFetch(
-		// If request doesn't contain an authenticator, continue has normal
+		// If request doesn't contain an authenticator, continue as normal
+		// Default to finCsrfFetch as fetch method has not been passed
 		resp = await finCsrfFetch(resource, options)
```
You're still calling `finCsrfFetch`. You need to call `fetchMethod` instead.

EliteScientist replied:
Yes. In this case we do not have a fetch method. The fetch method is passed via the options. this case is when there are no options or the options are a RequestInit that does not extend RequestInitAuth; where the fetchMethod is implemented by RequestInitAuth.

garethj2 commented on src/client/finAuthFetch.ts:
```diff
@@ -91,7 +91,24 @@ export async function finAuthFetch(
+	 * @param options Optional object containing any custom settings.
+	 * @returns A promise that resolves to a response object.
+	 */
+	fetchMethod?: (
```
Following the same design for `Client.ts`, please use `fetch` instead of `fetchMethod`.

Also see how `fetch` is defined in `Client.ts`...

    import { FetchMethod} from './fetchVal'
    
    fetch?: FetchMethod

Commit:
PR Changes:
 - Use same design as Client.ts for fetch
